### ISSUE TITLE
User scripts component

### ISSUE THREAD

**Issue by uazo** [bromite/bromite#792]
### Is your feature request related to privacy?

No, but it could become

### Is there a patch available for this feature somewhere?

No.

### Describe the solution you would like

as some users have said, I also find privacy pop-ups boring, especially google ones, since in my phone they are quite lagged

### Describe alternatives you have considered

initially I thought about reusing the Inject-scripts-for-AMP-tracking-ads-and-video.patch but then I saw that there is the possibility to activate the support for userscripts, which is somehow better because it leaves no trace in the dom, as opposed to that patch.

as a personal test (and also a bit of a challenge) I tried to reactivate user scripts in bromite (https://www.chromium.org/developers/design-documents/user-scripts) taking them from the extensions and moving them to a new one component, managing to make them work (https://github.com/uazo/bromite/blob/gitpages/userscripts_v2.7z my work)
Obviously, as my first test, there is no ui (and by the way it could be quite long to implement it), we could think to proceed in steps.

I wanted to understand if there was interest and, if so, if we wanted to discuss how to implement it, also considering the security aspects, updates and possible integration with chromium developer tools.

current state:
- [x] base code + reading from folder
  - [x] add details for unmatch header
  - [x] bug DOCUMENT_START not work
- [x] preferences (enable/disable)
  - [x] add link to help page (actually https://github.com/bromite/bromite/wiki/UserScripts)
- [x] ui management (enable/disable/remove script) 
  - [x] consider changing proguard settings for R (see https://stackoverflow.com/a/2889844)
  - [x] add icons (enable/disable)
  - [x] disable add button if feature is disabled
  - [x] bug enable/disable need restart
- [x] add script from file dialog
  - [x] remove excessive permissions prompt from filedialog (camera + audio)
  - [x] check how to transform a content:// (URI) to file:// (PATH)
  - [x] get real script filename from a content://
  - [x] when importing from file, skip non ascii chars used in description to prevent [FATAL:web_string.cc(112)] Check failed: ContainsOnlyASCII()
  - [x] reload list after add and show toast with error (check how works CallByNative)
  - [x] check if a file is real user script (es refuse txt without javascript)
- [x] understand if installation from download makes sense (entry point https://source.chromium.org/chromium/chromium/src/+/master:chrome/android/java/src/org/chromium/chrome/browser/download/DownloadUtils.java;l=360), actually we have to download the file
  - [x] check how to request permission from the user
  - [x] not working
- [x] add feature flag
- [x] consider if we want to add a way to prevent crash in the startup
- [ ] code cleaning
  - [x] remove unused code
  - [x] remove unused BrowserContextHandle from bridge
  - [ ] try to simply rendering code
  - [x] change all ifdef removing extension word
  - [x] add a flag for logging
  - [x] add file name to pickle (file) so we can log script name at injection time
  - [x] check compiling webview
- [ ] ui management (2)
  - [ ] script details
  - [ ] disable in incognito
  - [x] view source script
  - [ ] check if we can use an "interstitial" page to install a userscript
  - [ ] localize error strings
- [ ] security verification
  - [ ] consider whether to use a sha256 to verify a possible modification of the file without the user's knowledge (if we give the possibility to choose the scripts folder)
  - [x] check if we can deny network access to injected scripts (check ScriptContext)
  - [x] permit only content:// from pickerui
  - [x] block install intent from outside --> no, ask user
- [ ] understand if it makes sense to use zip files (for adding css directly, userscripts.js doesn't support them directly) code https://source.chromium.org/chromium/chromium/src/+/master:components/services/unzip/public/cpp/unzip.cc
- [ ] understand if self-updating makes sense
  - [ ] check how the source can be verified (script is unsigned by default)
  - [ ] or enable include metadata support
- [ ] add management in devtools (like snippets, code https://source.chromium.org/chromium/chromium/src/+/master:third_party/devtools-frontend/src/front_end/snippets/ScriptSnippetFileSystem.js)
  - [ ] add flag for "fall-back" in devtools (code https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/android/devtools_server.cc)
- [ ] add test cases
- [x] try with v88
- [x] try with v89
- [x] try with v90
- [x] try with v91
- [x] bug  [FATAL:utf_string_conversions.cc(338)] Check failed: IsStringASCII(utf16). Tampermonkey • Userscript Sources
- [x] handling unicode scripts

**Comment by csagan5** [comment]
> I tried to reactivate user scripts
> taking them from the extensions and moving them to a new component

What is the status of this feature upstream? is `--enable-user-scripts` still working?

**Comment by uazo** [comment]
That switch was removed on 2009-11-24
https://source.chromium.org/chromium/chromium/src/+/faed6e1304f7ba31ed12d42ea3dc694021690afd

but that code is alive (see for example https://source.chromium.org/chromium/chromium/src/+/6d96d3d1cb83d433c2b6a7c2f29c15900296f3c8)
since it is used for declarative scripts used by extensions

**Comment by csagan5** [comment]
How would this work on Android? I have checked your 7z file but since it is not a patch I cannot see exactly what you changed without extracting it on a branch.

> possible integration with chromium developer tools.

What integration do you mean here?

**Comment by uazo** [comment]
> I cannot see exactly what you changed without extracting it on a branch.

it's because I haven't changed anything, it's all new, taking code from extensions and removing the superfluous, they are not active in android.
it is a new component, like those already found in the base code. I preferred this way, given the need to keep maintenance easy, it is something we can discuss.

> How would this work on Android?

technically the component attaches itself to the rendering process (it is the default behavior of any existing component) and, responding to events in chromium, injects javascript code declaring when and what.
the browser process, via mojo, sends the scripts to the rendering engine, since it cannot access the disk to retrieve any scripts.
all things I've already found, of course, credit to google teams.

Of all this, I haven't done this last thing yet, but the code already provides for it. for now i just statically declared the scripts, you can see it in `components\user_scripts\content\renderer\user_script_set_manager.cc`

What can be injected is visible here https://source.chromium.org/chromium/chromium/src/+/master:extensions/common/user_script.h while here https://source.chromium.org/chromium/chromium/src/+/master:extensions/browser/extension_user_script_loader.cc and here https://source.chromium.org/chromium/chromium/src/+/master:extensions/common/manifest_handlers/content_scripts_handler.cc the classes that loads from file.

> What integration do you mean here?

I mean that it would also be possible to think of adding the control of user scripts through the chromium developer tools, to make their development easier. this is just an idea, I don't know how to do it yet.

**Comment by csagan5** [comment]
It is a quite big change, almost comparable to extensions support. I would have to rely on your to maintain it in future versions if it were to be merged.

I understand the integration with developer tools (I guess you mean via remote debugging), but what about Android regular users? It would need some UI to load/activate/disable such user scripts?

**Comment by uazo** [comment]
> I guess you mean via remote debugging

yes, I mean that.

> but what about Android regular users? It would need some UI to load/activate/disable such user scripts?

yes, they would have to be developed.
in the next few days I will prepare a proposal to discuss.

> I would have to rely on your to maintain it in future versions if it were to be merged.

let's talk about. my idea is to minimize the interventions: developing a new component, the maintenance is the opposite of what you are used to, so it is not a patch from the chromium code, but it is a patch to the chromium code.
I hope I have explained myself, the language does not help me.

**Comment by csagan5** [comment]
I understand the idea of a component; still, you will see that even with components maintenance is a burden.

**Comment by uazo** [comment]
> you can already create a PR rather than a 7z archive?

you can see it here.

https://github.com/uazo/bromite/blob/enable-user-scripts/build/patches/add-user-scripts.patch

sorry for the delay, but I had the branch blocked by the other patches.

**Comment by csagan5** [comment]
> you can see it here.
> 
> https://github.com/uazo/bromite/blob/enable-user-scripts/build/patches/add-user-scripts.patch

I have not had time to test this yet, but what are the incomplete parts you can individuate?

**Comment by DI555** [comment]
@uazo , could you put there support and for userstyles , please ?

**Comment by uazo** [comment]
version 3 [wip]:
https://github.com/uazo/bromite/blob/112c6d8aa511f2d6b50245374fc127ea7c777d1b/build/patches/wip-user-scripts-v3.patch

Added:
- use of folder `snippets` for userscripts under app folder

currently for all files in that folder parse and use them if they are correct. any error messages are in the log.

what is missing:
- ui
- security verification: consider whether to use a sha256 to verify a possible modification of the file without the user's knowledge
- understand if it makes sense to use zip files (for adding css, .userscripts.js doesn't support them)
- understand if installation from download is possible
- understand if self-updating makes sense
- add management in devtools
- code cleaning

**Comment by csagan5** [comment]
> * security verification: consider whether to use a sha256 to verify a possible modification of the file without the user's knowledge

This is not necessary: Android has sandboxing, no other app can modify another app's files.

> * understand if installation from download is possible

The AdBlock updater in Bromite already stores files locally, it could be used as a reference.

> * understand if self-updating makes sense

This should be considered only later on, when there is an UI.

> * ui

This might take a lot of time, I would look at it last. We still do not have #811

**Comment by uazo** [comment]
> This is not necessary: Android has sandboxing, no other app can modify another app's files.

yes, but I thought so in case we wanted to leave the user the possibility to choose the scripts folder, so that eventually he could choose to do a sync

> The AdBlock updater in Bromite already stores files locally, it could be used as a reference.

not quite like that, I wanted to use this
https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/download/chrome_download_manager_delegate.cc;l=659

> This should be considered only later on, when there is an UI.

Sure

> This might take a lot of time, I would look at it last. We still do not have #811

uhm, I'm working on it now because it's a key part. it is not very long (with cut&paste cr code), of course if each test did not take 10 minutes between compiling and installing it would be better! let's see, a new pc is coming.

it's the tests that worry me more and memory management for any leaks

**Comment by csagan5** [comment]
> so that eventually he could choose to do a sync

I do not think that it is possible. Android 11 implements even stronger sandboxing between apps/OS.

**Comment by csagan5** [comment]
@uazo is this related/useful? https://bugs.chromium.org/p/chromium/issues/detail?id=1054624

**Comment by uazo** [comment]
>  is this related/useful? https://bugs.chromium.org/p/chromium/issues/detail?id=1054624

not currently, for what I'm doing. I don't directly use the api of the extensions but only their logic.
But it's interesting how you got there: is there any way to register for notifications to a specific bug area? it would also be useful if it were by file or by directory.

in the meantime I ask you: I just can't fully understand the logic behind of `base::BindOnce()` that is driving me crazy, any suggestions of where to look for good documentation? I didn't find much in the code.

Anoher thing, would you be inclined to modify the proguard settings excluding R?
if, instead of changing the preferences every time directly in the xml, we change `SettingsUtils.addPreferencesFromResource()` and through the reflections we search for all the xml in R that ended in ".addon.xml", we would not have overlaps between different patches.
it's just an idea but in debug (without proguard) i tried and it works: please wait for me to put up the updated patch so I can show you the code (as soon as I understand how the hell it works BindOnce!)

**Comment by csagan5** [comment]
> But it's interesting how you got there: is there any way to register for notifications to a specific bug area? it would also be useful if it were by file or by directory.

I found this one while looking for something else; you can create saved queries in the Chromium issue tracker and get email notifications for them.



> in the meantime I ask you: I just can't fully understand the logic behind of `base::BindOnce()` that is driving me crazy, any suggestions of where to look for good documentation? I didn't find much in the code.

I am not an expert but it seems to be a way to invoke a function only once, based on its pointer. See
https://github.com/chromium/chromium/blob/87.0.4280.106/base/callback.h#L30 and then below there is the documentation for `Then`:
```
  // Then() returns a new OnceCallback that receives the same arguments as
  // |this|, and with the return type of |then|. The returned callback will:
  // 1) Run the functor currently bound to |this| callback.
  // 2) Run the |then| callback with the result from step 1 as its single
  //    argument.
  // 3) Return the value from running the |then| callback.
  //
  // Since this method generates a callback that is a replacement for `this`,
  // `this` will be consumed and reset to a null callback to ensure the
  // originally-bound functor can be run at most once.
```



> Anoher thing, would you be inclined to modify the proguard settings excluding R?

You mean something like this? https://stackoverflow.com/a/2889844

> instead of changing the preferences every time directly in the xml, we change `SettingsUtils.addPreferencesFromResource()` and through the reflections we search for all the xml in R that ended in ".addon.xml", we would not have overlaps between different patches.

It will bloat a bit the APK, but that is not a big concern. Do we have many XML issues with patches?



> it's just an idea but in debug (without proguard) i tried and it works: please wait for me to put up the updated patch so I can show you the code (as soon as I understand how the hell it works BindOnce!)

Ok.

**Comment by uazo** [comment]
some screenshot:

| **Settings:**  | **Enable/Disable script:** |
| ------------- | ------------- |
| ![image](https://user-images.githubusercontent.com/29201891/101983252-1cb17280-3c7a-11eb-953f-03f0b8f85794.png)  |  ![image](https://user-images.githubusercontent.com/29201891/101983268-44083f80-3c7a-11eb-9ca2-8885136f4e18.png) |

**File Picker:**
![image](https://user-images.githubusercontent.com/29201891/101983257-32269c80-3c7a-11eb-9d2f-43f0a2074388.png)

**Sample for https://greasyfork.org/it/scripts/34613-youtube-multi-downloader-v6-0-youtubeconverter-io-mp3-fhd-mp4-hd-sd-m4a-3gp:**
![image](https://user-images.githubusercontent.com/29201891/101983318-a6614000-3c7a-11eb-9154-877a184595fd.png)

**Comment by uazo** [comment]
> I found this one while looking for something else; you can create saved queries in the Chromium issue tracker and get email notifications for them.

thanks, I'll try.

> be a way to invoke a function only once, based on its pointer

yes, what I didn't understand is how. for now I found that it actually creates a function that allows the regeneration of parameters within the other thread: what I did not understand in fact is why some types of parameters accepted them and others did not, and I solved it by understanding that certain parameters can be recreated only if defined as constants. here i miss the basics of c ++, it has changed a lot since my university days.

> You mean something like this? https://stackoverflow.com/a/2889844

Yes perfect! I just have to figure out where it is in the source ...

> Do we have many XML issues with patches?

actually and just to simplify switching between different releases. the fewer interventions I can do in the source code base, the better, right?

> instead of changing the preferences every time directly in the xml, we change SettingsUtils.addPreferencesFromResource()

See here:
https://github.com/uazo/bromite/blob/f6d9f2ebcc384c3e75d921fbb3f3bb309122096e/build/patches/add-user-scripts.v4.patch#L230
https://github.com/uazo/bromite/blob/f6d9f2ebcc384c3e75d921fbb3f3bb309122096e/build/patches/add-user-scripts.v4.patch#L435
https://github.com/uazo/bromite/blob/f6d9f2ebcc384c3e75d921fbb3f3bb309122096e/build/patches/add-user-scripts.v4.patch#L495

**Comment by csagan5** [comment]
> what I did not understand in fact is why some types of parameters accepted them and others did not

They are constraints created by the template. Since the function is run by another thread only some operations are allowed, to not have memory leaks or invalid access between threads.



> actually and just to simplify switching between different releases. the fewer interventions I can do in the source code base, the better, right?
> See here:
> https://github.com/uazo/bromite/blob/f6d9f2ebcc384c3e75d921fbb3f3bb309122096e/build/patches/add-user-scripts.v4.patch#L230
> https://github.com/uazo/bromite/blob/f6d9f2ebcc384c3e75d921fbb3f3bb309122096e/build/patches/add-user-scripts.v4.patch#L435
> https://github.com/uazo/bromite/blob/f6d9f2ebcc384c3e75d921fbb3f3bb309122096e/build/patches/add-user-scripts.v4.patch#L495

There are pros/cons; the addon XML would also have linting checks applied? About missing resources etc

**Comment by uazo** [comment]
> the addon XML would also have linting checks applied?

Yes

**Comment by kokukyocho** [comment]
Hello @uazo!

I am a bot :robot:

Thanks for submitting this issue!
Sorry, this is not a feature request related to privacy. It might be considered for development but it is invalid because - as explained in the template - only feature requests related to privacy are accepted.

**Comment by kokukyocho** [comment]
Hello @uazo!

I am a bot :robot:

Thanks for submitting this issue!
Sorry, this is not a feature request related to privacy. It might be considered for development but it is invalid because - as explained in the template - only feature requests related to privacy are accepted.

**Comment by kokukyocho** [comment]
Hello @uazo!

I am a bot :robot:

Thanks for submitting this issue!
Sorry, this is not a feature request related to privacy. It might be considered for development but it is invalid because - as explained in the template - only feature requests related to privacy are accepted.

**Comment by kokukyocho** [comment]
Hello @uazo!

I am a bot :robot:

Thanks for submitting this issue!
Sorry, this is not a feature request related to privacy. It might be considered for development but it is invalid because - as explained in the template - only feature requests related to privacy are accepted.

**Comment by kokukyocho** [comment]
Hello @uazo!

I am a bot :robot:

Thanks for submitting this issue!
Sorry, this is not a feature request related to privacy. It might be considered for development but it is invalid because - as explained in the template - only feature requests related to privacy are accepted.

**Comment by DI555** [comment]
@uazo , many thanks for keeping work on it!!!
and you meaned support for UserCSS, we're looking for this much also!!!

**Comment by uazo** [comment]
> many thanks for keeping work on it!!!

thanks, sooner or later it will see the light!

> and you meaned support for UserCSS

the function is already present, consider that a userscript can easily inject all the css it wants

**Comment by DI555** [comment]
@uazo , yes, i know about that..
but, may be rightness and secure, if .css will be loaded more natural way, without using the script for that ?

**Comment by DI555** [comment]
@uazo , could you make user styles (user css) work out-of-the-box either please ? ..injection each time it into user.js is very uncomfortable((

**Comment by uazo** [comment]
I had searched long ago for specifics to follow, finding nothing.
do you have any sites that talk about how a "usercss" file (which is not an extension) is made?

**Comment by DI555** [comment]
it seems there're two file names, providing user .css :
.user.css (stylish one) and .user.styl (stylus one)

basics here:
github.com/openstyles/stylus/wiki/UserCSS

more detail info here:
github.com/openstyles/stylus/wiki/Writing-UserCSS

**Comment by csagan5** [comment]
This feature is left open to implement the remainder described in the issue description; @uazo there is no intent to develop anything related to UserCSS and please do not.

@DI555 you can use the existing functionality of user scripts that works only with Javascript, we will not change Bromite to support UserCSS and then chase the infinite bugs due to that.

**Comment by quantumleap-** [comment]
This is a great useful feature.
However, at present, it seems that multibyte characters in the code are replaced with blanks.

**Comment by uazo** [comment]
> it seems that multibyte characters in the code are replaced with blanks.

could you give us some examples?

**Comment by quantumleap-** [comment]
@uezo Thank you for your reply.  
The issue of multibyte characters being replaced by blanks is described below with sample code.

- The "name" and "description" of the script displayed in the script management page will be displayed correctly, including multibyte characters.
- In the script code displayed by "View source" from the script management page, multibyte characters will be blank.
- In the script that is actually executed, multibyte characters will be blank.

```
// ==UserScript==
// @name ⚠test⚠ multibyte characters display.
// @version 0.1.00
// @description ⚠test⚠ en: This is test script. Bromite is a great FLOSS product 👍/ja :これはテスト用スクリプトです。ブロマイトは素晴らしいFLOSS製品です 👍
// @run-at document-end
// ==/UserScript==
'use strict';
(() => {
  alert('English: [This is test script. Bromite is a great FLOSS product 👍]\nJapanese: [これはテスト用スクリプトです。ブロマイトは素晴らしいFLOSS製品です 👍]\n');
  document.body.insertAdjacentHTML('beforeend','<div id="test-message-container" style="position:fixed;top:0;height:50px;width:100vw;background:#e33a;display:grid;place-items:center;justify-content:center;z-index:10000;"><p>Thumbs Up Emoji:[👍]</p></div>');
}
)();
```

**Comment by csagan5** [comment]
I wonder whether it is a problem of the file being downloaded and decoded/re-encoded before writing with incorrect encoding, or loaded with incorrect encoding as an user script.

**Comment by uazo** [comment]
instead I seem to remember doing it on purpose, out of laziness :) it's been a while...
because managing unicode strings from native code seemed complex to me. but only lately in chromium I have seen that something has changed about it, but I have not checked yet ...

**Comment by Disqu3-mirOir-qt** [comment]
for me i test my script and not work with bromite but work with other software support userscript

![Screenshot_20211202-161304.jpg](https://user-images.githubusercontent.com/41585567/144449394-fbaba8df-9306-46a3-83de-26336e751236.jpg)

![Screenshot_20211202-161256.jpg](https://user-images.githubusercontent.com/41585567/144449419-db83b77a-9098-400e-bf7e-78e9938edb48.jpg)


![Screenshot_20211202-161248.jpg](https://user-images.githubusercontent.com/41585567/144449440-a11a2831-710c-47df-bd37-7784e69fddf6.jpg)

**Comment by uazo** [comment]
@Disqu3-mirOir-qt is correct, content scripts do not run in the scope of the current page (see https://developer.chrome.com/docs/extensions/mv3/content_scripts/#isolated_world)
try using the run-at clause as described [here](https://github.com/bromite/bromite/wiki/UserScripts)
unlike other browsers, userscripts in bromite are more similar to chrome extensions, if you search on the internet you will find all the details

**Comment by csagan5** [comment]
> instead I seem to remember doing it on purpose, out of laziness :) it's been a while...
> because managing unicode strings from native code seemed complex to me. but only lately in chromium I have seen that something has changed about it, but I have not checked yet ...

Please add it to the issue TODOs then?

**Comment by xarantolus** [comment]
I'm not 100% sure about this, but it seems like only one script is run when multiple ones would match.

E.g. add the [I don't care about cookies](https://github.com/xarantolus/bromite-userscripts/blob/main/block/idcac.user.js?raw=true) and [Generalized Cookie Prompt Accept/Block](https://github.com/xarantolus/bromite-userscripts/blob/main/block/accept-cookie-prompts.user.js?raw=true) from [this repo](https://github.com/xarantolus/bromite-userscripts/) to Bromite. Then visit any page (e.g. google.com). Both scripts would log output (containing the script names) when loaded, but only one script seems to run/logs to the console. Now when I go to settings and disable the script that just ran, then go back to the page and reload it, the other script will run.
So to me this looks like both scripts work individually, but when both are enabled, only one script runs. The same is also true for pages that match even more scripts, only one script gets executed.

**Comment by quantumleap-** [comment]
＠xarantolus 
I have confirmed the same behavior.

Also, when multiple scripts with different execution timings (@run-at) are enabled, the script that is executed first seems to be executed at the remaining timings.

Steps to Reproduce:
Enable only the three test scripts shown [here](https://zerobin.net/?7c249f874aed15a4#5d0Iqp40dqFMn70SZah/f9h/63krVVnx+7XCWn/tvN4=) and reload the page. We should see "alert 1" appear three times (at document-start and document-end and document-idle).

**Comment by uazo** [comment]
ok, thank you all, I will check and let you know.

**Comment by ncaat** [comment]
It's not that important but it would be nice to have a flag to change the default CSP set in line 10497 of the userscript patch. For some of my userscripts to work I had to recompile bromite with a different CSP allowing eval.

**Comment by uazo** [comment]
I fixed all the problems, soon I send a pull.

> it would be nice to have a flag to change the default CSP ...  with a different CSP allowing eval.

I currently use this `script-src 'self'; object-src 'self'` as indicated https://source.chromium.org/chromium/chromium/src/+/main:extensions/common/manifest_handlers/csp_info.cc;l=66

```
// The default CSP to be used in order to prevent remote scripts.
```

I don't know if it is appropriate to also add `unsafe-eval`

@ncaat you can try to check if it works for you https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function#difference_between_function_constructor_and_function_declaration ?

**Comment by ncaat** [comment]
@uazo without `'unsafe-eval'` all methods that create code from strings are blocked: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src#unsafe_eval_expressions
I think it's already sandboxed so maybe it's fine to add `'unsafe-eval'`? I'm not sure though. Regardless it would still be nice to have a flag in chrome://flags to be able to change the CSP without recompiling.

**Comment by uazo** [comment]
> all methods that create code from strings are blocked

ah, ok.

> I think it's already sandboxed so maybe it's fine to add 'unsafe-eval'? Regardless it would still be nice to have a flag in chrome://flags to be able to change the CSP without recompiling.

it is not possible to manually change strings, since `chrome://flags` does not allow user input other than urls.
but I can insert an on/off flag that activates the evals, @csagan5 do you agree? the current extension code uses the page csp, that was my addition

**Comment by csagan5** [comment]
> I can insert an on/off flag that activates the evals, @csagan5 do you agree? the current extension code uses the page csp, that was my addition

It is not a trivial difference; with a different CSP that means you allow any code dynamically generated by a domain controlled by the author of the script to end up running in your webpages. It would be perfect to steal credentials; while instead with the current CSP only the code in the `user.js` file is the one that runs.

If this option is introduced then we might end up with most of the scripts being created needing `unsafe-eval`, leading to a generally relaxed security measure. What would be the response when people start to get hacked this way?

Perhaps if you can think of a way to grant it to specific scripts, that would be better; but the observation above still applies: my offering an option to remove the sandbox we are encouraging lax security.

**Comment by ncaat** [comment]
@csagan5 any userscript can steal your data. You definitely shouldn't run any scripts that you haven't either written yourself or thoroughly checked. The flag would also be deactivated by default. My script uses eval() to execute user input as part of a javascript console for mobile. To still benefit from the better security I could deactivate the flag after I'm done. There aren't a lot of things that need eval to work so I don't think many scripts would use it. I think the PR is fine as it is.

**Comment by csagan5** [comment]
@ncaat I have already explained the difference, see previous message. We are not going to have a flag to enable `eval()` for all scripts.

**Comment by uazo** [comment]
> It is not a trivial difference; with a different CSP that means you allow any code dynamically generated by a domain controlled by the author of the script to end up running in your webpages

okay, i will remove the flag, i trust your views on safety.

> Perhaps if you can think of a way to grant it to specific scripts, that would be better;

I will do so.

**Comment by uazo** [comment]
@ncaat I added the flag to allow evals, manageable from the interface.
in the meantime you can find it here https://github.com/uazo/bromite/blob/master-v97/build/patches/Experimental-user-scripts-support.patch

**Comment by csagan5** [comment]
> @csagan5 any userscript can steal your data. You definitely shouldn't run any scripts that you haven't either written yourself or thoroughly checked.

The `eval` capability is to allow a script which you have not yet read, nor possibly can, because it can be sourced from a third-party website. This is a substantial difference.

Current scenario: you can read the script that you are going to execute, fully.
Scenario with eval: you never know what script is going to be executed.

So, this is not the same: it is exactly the opposite security-wise and the equivalent [ActiveX](https://en.wikipedia.org/wiki/ActiveX) on Android.

@uazo since it was not agreed  to add this feature I will not add it for v97 and it is not sure that we will ever add it, I am sorry that you did all the work already before a decision was taken here.

**Comment by uazo** [comment]
> since it was not agreed to add this feature I will not add it for v97 and it is not sure that we will ever add it

no problem, I figured having a flag for userscript was fine. 

> Scenario with eval: you never know what script is going to be executed.

However, this may not be true, having evals active does not necessarily mean insecure...

> because it can be sourced from a third-party website.

How? userscript are isolated should not have the ability to exchange data with other scripts and I could deny access to the network with `connect-src` csp.

**Comment by ncaat** [comment]
@csagan5 while I agree that it can impact security so can simply having javascript enabled. It's often a trade-off between functionality and security. IMO the decision should be made by the users to decide for themselves whether they think it's worth it. Adding the option to enable eval with it being disabled by default doesn't impact the security of someone who doesn't activate it.

**Comment by csagan5** [comment]
@ncaat I have already explained why allowing such functionality creates a slippery slope; feel free to study how `eval()` is allowed (or not) in GreaseMonkey, TamperMonkey etc

@uazo we should close this issue once all the tasks in the issue description are addressed, so that separate issues can be created over time (and not discussed here, which makes it extremely difficult to follow and to reference in future)

**Comment by ncaat** [comment]
@csagan5 GreaseMonkey and TamperMonkey scripts can even disable their sandbox.

**Comment by uazo** [comment]
> UserScripts in Bromite will not support unsafe-eval.

however, I think I will do so.
either I make two different versions or if for @csagan5 is good for bromite, it will not be possible to activate it in any way because the code will not be included in the build through a gn flags.

**Comment by csagan5** [comment]
> however, I think I will do so.

Anyone can compile any Chromium variant with any patches, in fact I would like to encourage more people to do so; but it will not be Bromite.

@uazo rather than GN flags I have been thinking to add other patches under `patches/contrib`; please make a PR to add it there on top of v97. This way we have the official Bromite patches and any other patch there, separately. The bottom bar or even @nikolowry's patch could be there.

**Comment by uazo** [comment]
> I have been thinking to add other patches under patches/contrib

uhm.. don't be offended but why should I put patches that bromite doesn't want in the bromite repo? I would be too tied to your times (and I repeat, do not be offended). perhaps a possible link in the readme is better, or another repo a little more open: among other things, this solution would allow, by acting on the deps, to combine it with the build process.

> rather than GN flags

can you explain me better? the official bromite patch and a patch of the patch? it seems to me more complex to maintain than a flag

**Comment by csagan5** [comment]
Feel free to proceed as you like; I will not accept patches with GN flags for this.

> why should I put patches that bromite doesn't want in the bromite repo?

For visibility, others building Bromite might be interested; `contrib` patches (intended as community contributions) is a well-known concept for Debian packages and similar.

**Comment by uazo** [comment]
fixed in [99.0.4844.55](https://github.com/bromite/bromite/commit/f645da9e44acdecb6040021b0428ff663fdda376)

**Comment by csagan5** [comment]
@uazo there are no other fixes needed for this? The issue has still some open TODOs.

**Comment by csagan5** [comment]
There are none opened before closing the issue?

**Comment by xarantolus** [comment]
I'm not entirely sure if this is an issue with the User scripts component itself, but it appears to only happen when it's enabled (and currently only with one script): when opening any page in Bromite, then switching to another app/the launcher, then switching back to Bromite, the page just freezes for 20+ seconds. To be more precise, only the displayed web page freezes, the home button and the URL bar continue to work.

I am pretty sure that the script does not contain any infinite loops, however it does contain quite large objects (several megabytes).

Steps to reproduce:
1. Enable *only* [this cosmetic filter script](https://github.com/bromite/bromite/files/8212564/cosmetic.user.zip) in User Script settings
2. Visit any webpage, the script is injected and works fine (it logs to the browser console)
3. Switch to another app (or the launcher)
4. Switch back to Bromite
5. The page is now frozen and doesn't work anymore.
     * The URL bar and home page button of Bromite still work
     * Tapping any element on the page (e.g. text inputs) will lead to the page being unfrozen (not sure if interaction is required for unfreezing), but only after a significant (more than 20 seconds) delay

I took a [performance profile](https://github.com/bromite/bromite/files/8212666/Profile-20220309T083923.zip), it shows a "Long task" taking 24 seconds while the page is frozen. So basically at around 5 seconds I switched back to Bromite and then it started freezing.

**Comment by uazo** [comment]
is reproducible, thanks for reporting. the behavior is similar to that described in #1837

**Comment by uazo** [comment]
@xarantolus could you tell me if you detect the same problem in incognito mode?

**Comment by xarantolus** [comment]
It seems like it does not happen at all in incognito mode, the script is injected and everything works fine.
In "normal mode" it also freezes when adding a new tab, but when adding a new incognito tab it does not freeze.

**Comment by uazo** [comment]
I figured out what that code is for.
substantially chromium, when the device goes offline, automatically uses the self-generated pages from the disk cache instead of simply indicating "No internet connection". removing that call, only bfcache is used and not the one on disk.

**Comment by csagan5** [comment]
> substantially chromium, when the device goes offline, automatically uses the self-generated pages from the disk cache instead of simply indicating "No internet connection". removing that call, only bfcache is used and not the one on disk.

One could check the commits of when this was introduced to learn more.
Few questions:
* does it work equally when it is removed? (if page still has all the pieces in the `bfcache`)
* is this the same as #1837? I assume not
* you would remove this also for normal mode to fix this specific UserScripts issue?

Perhaps it is intended as a privacy feature because the `bfcache` will not contain the 'No-Cache' resources, like trackers, which would then be fetched again when accessing recents in order to be able to render a thumbnail.

**Comment by uazo** [comment]
> One could check the commits of when this was introduced to learn more.

yes, seen from there and from open bugs

> does it work equally when it is removed? (if page still has all the pieces in the bfcache)

yes, if page still has all the pieces in the bfcache but the bfcache is more limited than the disk cache and may not be active depending on the amount of memory in the device.

> is this the same as https://github.com/bromite/bromite/issues/1837? I assume not

I hoped so, but no, I have already installed on my device the version without that call and the freeze indicated by that issue happens the same.

> you would remove this also for normal mode to fix this specific UserScripts issue?

it depends on what we want to do, for me there is no problem in eliminating it.
actually the fundamental problem is that userscript, it shouldn't change the html that much if there is no need.
in fact the cache saves the html not the script, and the script modifies the page heavily.

> Perhaps it is intended as a privacy feature

I do not think so.
you know, for me the bfcache has some privacy problems, it would have to be checked, I don't have the elements yet to be able to say for sure. fundamentally it seems to me that it keeps in memory the state of the renderprocess when the navigation is cross domain: my concern is that the state of the javascript variables is also kept making it useless, but only for example, to delete cookies intra-session if the value is kept in a js variable.
this would render extensions like autodelete cookies (which I have installed on my desktop) useless. think about a search: I click on a site, cookie autodelete deletes the cookies of the search provider but when I go back, the search engine does not use the cookie but a variable and here it has understood who I am anyway.

**Comment by csagan5** [comment]
> > you would remove this also for normal mode to fix this specific UserScripts issue?
> 
> it depends on what we want to do, for me there is no problem in eliminating it.

Considering that we do not fully understand why it is there, I would not remove it until we do. Perhaps it is a mechanism to avoid bloating the `bfcache`. And since this is a specific example of an user script changing the DOM/HTML a lot, I would not change anything yet.

> for me the bfcache has some privacy problems, it would have to be checked, I don't have the elements yet to be able to say for sure. fundamentally it seems to me that it keeps in memory the state of the renderprocess when the navigation is cross domain: my concern is that the state of the javascript variables is also kept making it useless, but only for example, to delete cookies intra-session if the value is kept in a js variable.
> this would render extensions like autodelete cookies (which I have installed on my desktop) useless. think about a search: I click on a site, cookie autodelete deletes the cookies of the search provider but when I go back, the search engine does not use the cookie but a variable and here it has understood who I am anyway.

So the scenario you are explaining is:
1. visit page
2. have it go in the `bfcache` (visiting another page for example)
3. delete all cookies
4. press back
5. verify that a Javascript-stored value from the `bfcache` still contains the original cookie/session ID

I think this is working as intended because "clearing cookies" does not include "destroy my currently loaded pages". In this case the `bfcache` corresponds (roughly) to a page loaded in another tab: why would you want it to be destroyed when clearing cookies?
A "fix" could be to truncate the `bfcache` when cookies are erased, but that would lead to page reloading when you go back (or forward).

How does this behave in incognito? e.g. is the `bfcache` in incognito showing any different behavior? Probably not.

**Comment by uazo** [comment]
> So the scenario you are explaining is:
> visiting another page for example

bfcache works only for cross site navigation, i.e. it is used forwards or backwards only if the domain changes. browsing between the same site does not use bfcache, from what I understood

> why would you want it to be destroyed when clearing cookies?
> A "fix" could be to truncate the bfcache when cookies are erased

actually using cookie is the first thing that came to my mind, but the scenario you assume is correct

> How does this behave in incognito? e.g. is the bfcache in incognito showing any different behavior? Probably not.

I don't know yet

**Comment by zhmars** [comment]
@uazo Thanks for your work!

Some minor issues:
- When opening the script from `Downloads`, the user script will fail to add, it just show a error `Cannot open script source.` instead. see https://imgur.com/a/LGTWULz
- When selecting a file from the picker in the settings, instead show the file name on the confirm dialog, it show a pretty long string `content://com.android.providers.downloads.documents/xxx`. see https://imgur.com/a/FkXVjV2
- After I add the script finally, the file name is shown on the top of script description, but it is duplicate. It would be better to show the script name `Remove Google Consent` instead. see https://imgur.com/a/EDIYke6

**Comment by csagan5** [comment]
@uazo I will report here some things to check from the other "discussion", only the relevant technical part:
* the error message could be improved? by specifying the reason for failure to open script source
* SAF is mandatory since Android 11, correct? there is no other way for any app to request storage access with user interactions

**Comment by csagan5** [comment]
I will close this issue; @uazo please file new feature requests for the unchecked TODO items if you plan to add support for more features.
